packages/core/src/lib/history.unit.test.ts

```diff
@@ -0,0 +1,137 @@
+import { describe, expect, vi } from 'vitest';
+import { MINIMAL_CONFIG_MOCK } from '@code-pushup/test-utils';
+import { getCurrentBranchOrTag, safeCheckout } from '@code-pushup/utils';
+import { collectAndPersistReports } from './collect-and-persist';
+import { HistoryOptions, history, prepareHashes } from './history';
+import { upload } from './upload';
+
+vi.mock('@code-pushup/utils', async () => {
+  const utils: object = await vi.importActual('@code-pushup/utils');
+  let currentBranchOrTag = 'main';
+  return {
+    ...utils,
+    safeCheckout: vi.fn().mockImplementation((branch: string) => {
+      currentBranchOrTag = branch;
+    }),
+    getCurrentBranchOrTag: vi.fn().mockImplementation(() => currentBranchOrTag),
+  };
+});
+
+vi.mock('./collect-and-persist', () => ({
+  collectAndPersistReports: vi.fn(),
+}));
+
+vi.mock('./upload', () => ({
+  upload: vi.fn(),
+}));
+
+describe('history', () => {
+  it('should check out all passed commits and reset to initial branch or tag', async () => {
+    const historyOptions: HistoryOptions = {
+      ...(MINIMAL_CONFIG_MOCK as HistoryOptions),
+    };
+
+    await history(historyOptions, ['abc', 'def']);
+
+    expect(getCurrentBranchOrTag).toHaveBeenCalledTimes(1);
+    expect(getCurrentBranchOrTag).toHaveReturnedWith('main');
+
+    expect(safeCheckout).toHaveBeenCalledTimes(3);
+    // walk commit history
+    expect(safeCheckout).toHaveBeenNthCalledWith(1, 'abc', {
+      forceCleanStatus: undefined,
+    });
+    expect(safeCheckout).toHaveBeenNthCalledWith(2, 'def', {
+      forceCleanStatus: undefined,
+    });
+    // reset
+    expect(safeCheckout).toHaveBeenNthCalledWith(3, 'main', {
+      forceCleanStatus: undefined,
+    });
```

Comment:
The Vitest equality checks don't differentiate between missing props and props with `undefined` value.

```suggestion
    expect(safeCheckout).toHaveBeenNthCalledWith(1, 'abc', {});
    expect(safeCheckout).toHaveBeenNthCalledWith(2, 'def', {});
    // reset
    expect(safeCheckout).toHaveBeenNthCalledWith(3, 'main', {});
```